### PR TITLE
Python: ObjectAPI to ValueAPI: Foresight Additions

### PR DESCRIPTION
**Pull request by BekaValentine**
This PR is part of the effort to modernize the Expressions queries, moving away from the use of `refersTo` and associated APIs, and towards `pointsTo` and its associated APIs.

### Foresight Additions

There are some additions that are in-progress for foresight purposes. The old `refersTo` code makes extensive use of predicates like `theIntType()` which modernizes to `ClassValue::int_()`. Not all of the `the...Type()` predicates have been translated, presumably because they're being moved over on an as-needed basis. It seems useful to just move everything over since most of it is formulaic.

### CODE REVIEW

File: python/ql/src/semmle/python/objects/ObjectAPI.qll [hunk 1]
```diff
@@ -668,40 +668,57 @@ module ClassValue {
...
+    ClassValue tupleType() {
+        result = TBuiltinClassObject(Builtin::special("tuple"))
+    }
+    
+    /** Get the `ClassValue` for the `list` class. */
+    ClassValue listType() {
+        result = TBuiltinClassObject(Builtin::special("list"))
+    }
+    
+    /** The builtin class '(x)range' */
```
Comment (RasmusWL) [inline python/ql/src/semmle/python/objects/ObjectAPI.qll]:
```suggestion
    /** Get the `ClassValue` for `xrange` (Python 2), or `range` (only Python 3) */
```

File: python/ql/src/semmle/python/objects/ObjectAPI.qll [hunk 2]
```diff
@@ -668,40 +668,57 @@ module ClassValue {
...
+    ClassValue rangeType() {
+        result = TBuiltinClassObject(Builtin::special("xrange"))
```
Comment (RasmusWL) [inline python/ql/src/semmle/python/objects/ObjectAPI.qll]:
```suggestion
        major_version() = 2 and result = TBuiltinClassObject(Builtin::special("xrange"))
```

**Comment by BekaValentine** [comment]
I would like to propose actually that we move everything over to the `...Type()` naming convention to be consistent with the `the...Type()` predicates, and also so we don't have these strange problems with `int_` and `float_` etc. which are necessarily atypical b/c they conflict w/ QL keywords. I think it'd be better for everything to follow the same convention even if its more verbose b/c then people will be able to figure it out more easily, than to have special cases for the sake of brevity.

**Comment by tausbn** [comment]
I think for the moment, it's best to stick with the scheme we have. As you point out, if we were to change the API, we would have to deprecate the old predicates, wait a while, and then finally remove them. 

Also, I wonder if it would be more confusing to have all of these methods end in `Type`. This would mean we had `ClassValue::nonetypeType` (and `ClassValue::typeType`, even). Note also that we're still forced to keep the first character lowercase, lest the compiler complain.

We _should_ probably clean up in these predicates at some point in the future. I note, for instance, that at present it's `nonetype` and `staticmethod` but also `nameError`, so we're currently not consistent in how we style these (though one could argue that this is closer to what you would write in Python, within the constraints of the CodeQL lexer).

**Comment by BekaValentine** [comment]
I've swapped out the names, such that the predicate name is whats in the string-y name. Some of the pre-existing things are not consistent with this pattern tho, so we might want to push to that convention instead? I've also had to add _ to some of them to avoid conflicts w/ QL naming.

**Comment by tausbn** [comment]
I think my opinion on naming can be summed up as "as close to what you would get in Python as possible, lowercase the first letter if necessary, and add a `_` suffix if it clashes with a built-in". We don't _quite_ follow this consistently -- `builtinFunction` would have to be called either `builtin_function_or_method` (to match what `type(len)` returns) or `builtinFunctionType` (to match the `types` module).

Some of these I'm not sure we need at all, though, or will _ever_ need, for that matter. As far as I can see, `theInstanceType`, `theClassType`, and others are never used in our codebase, so there's no need to port them over. (This is in part the reason for why this was done in a piecemeal fashion up to now.)

(Note that some of them _are_ used elsewhere, like the `ObjectInternal classType()` predicate in `ObjectInternal.qll`, but this is fine. We're only going to deprecate `...Object`, not `...ObjectInternal`.)

Perhaps what we really need is a different interface for this. It seems to be that just piggy-backing onto the `types` module might be the right approach. That is, if people want `ClassType`, they can write `Value::named("ClassType")`. Or, at the very least, this can act as a way to access seldom-used types, while still keeping the common ones (e.g. `ClassValue::str`) close at hand.

(I note in passing that there may be a bug in how we calculate the result of `type` in our libraries. If you do `Value::named("types.InstanceType")`, it'll tell you it's a class named `_C`. This is the _class_ that's used in the `types` module to define `InstanceType`, but not `InstanceType` itself, which seems wrong.)

**Pull request merged**